File: CarRentalSystem/CarRentalSystemEntities/Customer.cs
using Newtonsoft.Json;
using System;
using System.Collections.Generic;

namespace sf.systems.rentals.cars
{
    public class Customer : ISerializedEntity<Customer>
    {
        public string Id { get; set; }
        public string Name { get; set; }
        public string PhoneNumber { get; set; }
        public string Address { get; set; }
        public string Email { get; set; }

        private readonly int CarRentalMaxLimit = 1;
        internal List<Car> RentedCarsPool { get; }

        public Customer() { }

        public Customer(string id, string name, string phoneNumber, string address, string email)
        {
            Id = id;
            Name = name;
            PhoneNumber = phoneNumber;
            Address = address;
            Email = email;
            RentedCarsPool = new List<Car>();
        }

        public List<Car> RentedCarsCopy
        {
            get
            {
                var result = new List<Car>();
                result.AddRange(RentedCarsPool);
                return result;
            }
        }

        internal void RentCar(Car car)
        {
            if (car == null) throw new ArgumentNullException("car");
            if (RentedCarsPool.Contains(car)) throw new InvalidOperationException($"Already rented car with ID: {car.Id}!");
            if (RentedCarsPool.Count >= CarRentalMaxLimit) throw new InvalidOperationException(
                $"Customer ({Id}) is not allowed to rent more than {CarRentalMaxLimit} car(s)!");

            car.Rent();
            RentedCarsPool.Add(car);
        }

        internal void ReturnCar(Car car)
        {
            if (RentedCarsPool.Contains(car)) RentedCarsPool.Remove(car);
            car.Return();
        }

        public string Serialize(DataType dataType) => dataType switch
        {
            DataType.CSV => $"{Id},{Name},{PhoneNumber},{Address},{Email}",
            DataType.JSON => JsonConvert.SerializeObject(this),
            _ => throw new ArgumentException($"Unknown data type {dataType}."),
        };

        public static Customer Deserialize(string data, DataType dataType) => dataType switch
        {
            DataType.CSV => CreateFromCsv(data),
            DataType.JSON => JsonConvert.DeserializeObject<Customer>(data),
            _ => throw new ArgumentException($"Unknown data type {dataType}."),
        };

        private static Customer CreateFromCsv(string csv)
        {
            string[] values = csv.Split(',');
            if (values.Length != 5)
            {
                throw new ArgumentException($"Invalid CSV data: {csv}");
            }
            return new Customer(values[0], values[1], values[2], values[3], values[4]);
        }

        public Customer DeserializeHandler(string data, DataType dataType)
        {
            return Deserialize(data, dataType);
        }

        public void RentedCarsPoolNew(List<Car> rentedCars)
        {
            RentedCarsPool.Clear();
            RentedCarsPoolExtend(rentedCars);
        }
        public void RentedCarsPoolExtend(List<Car> rentedCars)
        {
            if (rentedCars == null) throw new ArgumentNullException("rentedCars");

            RentedCarsPool.AddRange(rentedCars);
        }
    }
}

File: CarRentalSystemPython/entities/Customer.py
from __future__ import annotations
from typing import List, Type
import json

from ISerializedEntity import ISerializedEntity, DataType
from Car import Car
from Customer import Customer

class Customer(ISerializedEntity[Customer]):
    CAR_RENTAL_MAX_LIMIT = 1

    def __init__(self, id: str, name: str, phone_number: str, address: str, email: str):
        self._id = id
        self._name = name
        self._phone_number = phone_number
        self._address = address
        self._email = email
        self._rented_cars_pool = []

    @property
    def id(self) -> str:
        return self._id

    @property
    def name(self) -> str:
        return self._name

    @property
    def phone_number(self) -> str:
        return self._phone_number

    @property
    def address(self) -> str:
        return self._address

    @property
    def email(self) -> str:
        return self._email

    @property
    def rented_cars_copy(self) -> List[Car]:
        return self._rented_cars_pool.copy()

    def rent_car(self, car: Car) -> None:
        if car is None:
            raise ValueError("car cannot be None")
        if car in self._rented_cars_pool:
            raise ValueError(f"Already rented car with ID: {car.id}!")
        if len(self._rented_cars_pool) >= Customer.CAR_RENTAL_MAX_LIMIT:
            raise ValueError(f"Customer ({self._id}) is not allowed to rent more than {Customer.CAR_RENTAL_MAX_LIMIT} car(s)!")

        car.rent()
        self._rented_cars_pool.append(car)

    def return_car(self, car: Car) -> None:
        if car in self._rented_cars_pool:
            self._rented_cars_pool.remove(car)
        car.return_car()

    def serialize(self, data_type: DataType) -> str:
        if data_type == DataType.CSV:
            return f"{self._id},{self._name},{self._phone_number},{self._address},{self._email}"
        elif data_type == DataType.JSON:
            return json.dumps(self.__dict__)
        else:
            raise ValueError(f"Unknown data type {data_type}.")

    @staticmethod
    def deserialize(data: str, data_type: DataType) -> Customer:
        if data_type == DataType.CSV:
            return Customer.create_from_csv(data)
        elif data_type == DataType.JSON:
            return Customer(**json.loads(data))
        else:
            raise ValueError(f"Unknown data type {data_type}.")

    def deserialize_handler(self, data: str, data_type: DataType) -> Customer:
        return Customer.deserialize(data, data_type)

    @staticmethod
    def create_from_csv(csv: str) -> Customer:
        values = csv.split(",")
        if len(values) != 5:
            raise ValueError(f"Invalid CSV data: {csv}")
        return Customer(values[0], values[1], values[2], values[3], values[4])

    def rented_cars_pool_new(self, rented_cars: List[Car]) -> None:
        self._rented_cars_pool.clear()
        self._rented_cars_pool_extend(rented_cars)

    def rented_cars_pool_extend(self, rented_cars: List[Car]) -> None:
        if rented_cars is None:
            raise ValueError("rented_cars cannot be None")

